aten/src/ATen/core/dispatch/Dispatcher.h

```diff
@@ -325,10 +329,38 @@ template<class... Args> inline void unused_arg_(const Args&...) {}
 }
 
 template<class Return, class... Args>
-inline Return Dispatcher::callWithDispatchKey(const TypedOperatorHandle<Return(Args...)>& op, DispatchKey dispatchKey, Args... args) const {
+inline Return Dispatcher::callWithDispatchKey(const TypedOperatorHandle<Return(Args...)>& op, DispatchKey dispatchKey, bool useRecordFunction, Args... args) const {
   detail::unused_arg_(args...);  // workaround for a false-positive warning about unused parameters in gcc 5
   const KernelFunction& kernel = op.operatorIterator_->op.lookup(dispatchKey);
-  return kernel.template call<Return, Args...>(op, std::forward<Args>(args)...);
+
+  if (useRecordFunction && op.operatorIterator_->op.isObserved()) {
+    // Check if we need to run callbacks registered with RecordFunction
+    // If true and callbacks need inputs, we box the arguments and pass
+    // them into the callbacks and also into the kernel call
+
+    // Note: for perf reasons we wouldn't want to pass arguments into
+    // the function call or prematurely box them
+    at::RecordFunction guard(at::RecordScope::FUNCTION);
+    if (C10_UNLIKELY(guard.active)) {
```

Comment:
I think you can put the dispatch key filter test inside this `if`, and also move the `op.isObserved()` test down here.